shell/src/peer_manager.rs

```diff
@@ -1059,6 +1063,69 @@ fn resolve_dns_name_to_peer_address(
     Ok(addrs)
 }
 
+struct P2pPeersMap {
+    peers: HashMap<ActorUri, P2pPeerState>,
+    addr_set: HashSet<SocketAddr>,
+}
+
+impl P2pPeersMap {
+    fn new() -> Self {
+        Self {
+            peers: HashMap::new(),
+            addr_set: HashSet::new(),
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        debug_assert!(self.peers.is_empty() == self.addr_set.is_empty());
+        self.peers.is_empty()
+    }
+
+    fn len(&self) -> usize {
+        debug_assert!(self.peers.len() == self.addr_set.len());
+        self.peers.len()
+    }
+
+    fn try_insert(
+        &mut self,
+        peer_ref: PeerRef,
+        peer_address: SocketAddr,
+        bootstrap_requested_last: Option<Instant>,
+    ) -> Result<(), PeerManagerError> {
+        if self.addr_set.contains(&peer_address) {
```

Comment:
@akoptelov 
please check comment in jira TE-490 
we should better handle unique connections by public_key and not by socket_address, 
because we can run more nodes from one IP with different port settings lets say

so I would just extend `P2pPeerState` with public_key and check uniquess based on this attribute, so we probably dont need P2pPeersMap, but I like the impl. `try_insert`